plutus-core/plutus-ir/src/PlutusIR/Purity.hs

```diff
@@ -68,7 +68,8 @@ must be *conservative* (i.e. if you don't know, it's non-strict).
 
 -- | Will evaluating this term have side effects (looping or error)?
 -- This is slightly wider than the definition of a value, as
--- it includes things that can't be returned from the machine (as they'd be ill-scoped).
+-- it includes applications that are known to be pure, as well as
+-- things that can't be returned from the machine (as they'd be ill-scoped).
```

Comment:
@michaelpj I find this:

> as well as things that can't be returned from the machine (as they'd be ill-scoped).

very confusing. So for one, we don't have any machine for PIR, but I assume this talks about the UPLC subset of PIR. What's an example of something that can't be returned from the machine? A standalone variable I guess, since the comment talks about scoping, but it can't be returned not because it would be out of scope, but because the machine takes care of looking up variables and so we simply never run into handling variables when returning from the machine. In fact it is perfectly possible to return a free variable from the machine if you hide it behind a never-evaluated lambda.

I think we need a more formal definition of what "pure" means. Ideally, explicitly put into the context of our evaluation semantics (e.g. a variable wouldn't be pure if any of our evaluators was lazy). Maybe we can ask one of our formal methods guys to help out?

Reply:
> I find this ... very confusing

Me too. If I wrote that then that was not very helpful of past me :sweat_smile: 

I'd be very happy to have a clearer definition. I think I mean something like "when this term is evaluated, regardless of the context, will it have side effects or loop?" (note that we don't completely disregard the context, as you can tell it which variables are strict or not, as that tells you whether evaluating a variable would do anything). I _think_ that's fairly well-defined, but maybe I'm wrong. And it is true that it relies on an implicit model of PIR evaluation that it would be better to make explicit.

Reply:
I've created PLT-1448 for that.